### PR TITLE
feat: Add support for `ItemIndexRange` and `ItemIndexRangeHelper`

### PR DESCRIPTION
**Pull request by jeromelaban**
## PR Type

What kind of change does this PR introduce?

### <samp>🤖 Generated by Copilot at caba0b0</samp>

This pull request adds the `ItemIndexRange` and `ItemIndexRangeHelper` classes and their unit and integration tests. These classes are part of the Windows.UI.Xaml.Data namespace and are used by the `ItemsRepeater` control to manage ranges of items in the data source.

## PR Checklist

Please check if your PR fulfills the following requirements:

- [ ] Docs have been added/updated which fit [documentation template](https://github.com/unoplatform/uno/blob/master/doc/.feature-template.md) (for bug fixes / features)
- [ ] [Unit Tests and/or UI Tests](https://github.com/unoplatform/uno/blob/master/doc/articles/uno-development/working-with-the-samples-apps.md) for the changes have been added (for bug fixes / features) (if applicable)
- [ ] Validated PR `Screenshots Compare Test Run` results.
- [ ] Contains **NO** breaking changes
- [ ] Associated with an issue (GitHub or internal) and uses the [automatic close keywords](https://help.github.com/en/github/managing-your-work-on-github/linking-a-pull-request-to-an-issue).
- [ ] Commits must be following the [Conventional Commits](https://www.conventionalcommits.org/en/v1.0.0/#summary) specification.

## Other information

Internal Issue (If applicable):

### CODE REVIEW

File: src/Uno.UI/UI/Xaml/Data/ItemIndexRange/ItemIndexRange.mux.cs [hunk 1]
```diff
@@ -0,0 +1,156 @@
...
+		var length1 = range1.Length;
+
+		var firstIndex2 = range2.FirstIndex;
+		var length2 = range2.Length;
+
+		return ItemIndexRangeHelper.AreRangesEqual(firstIndex1, length1, firstIndex2, length2);
+	}
+
+	// return true if the two collection are identical (even in order of items)
+	// for example, {{5,1},{7,3}} is NOT equal to {{7,3},{5,1}}
```
Comment (agneszitte) [inline src/Uno.UI/UI/Xaml/Data/ItemIndexRange/ItemIndexRange.mux.cs:70]:
```suggestion
	// return true if the two collections are identical (even in order of items)
	// for example, {{5,1},{7,3}} is NOT equal to {{7,3},{5,1}}
```

File: src/Uno.UI/UI/Xaml/Data/ItemIndexRange/ItemIndexRange.mux.cs [hunk 2]
```diff
@@ -0,0 +1,156 @@
...
+		{
+			length = ItemIndexRangeHelper.GetContinousIndicesLengthStartingAtIndex(indices, i);
+
+			ItemIndexRange.AppendItemIndexRangeToItemIndexRangeCollection((int)indices[i], length, ref pCollection);
+		}
+
+		return pCollection;
+	}
+
+	// Used to create and ItemIndexRange and append it to the provided collection
```
Comment (agneszitte) [inline src/Uno.UI/UI/Xaml/Data/ItemIndexRange/ItemIndexRange.mux.cs:128]:
```suggestion
	// Used to create an ItemIndexRange and append it to the provided collection
```

File: src/Uno.UI/UI/Xaml/Data/ItemIndexRange/ItemIndexRangeHelper.h.mux.cs [hunk 3]
```diff
@@ -0,0 +1,39 @@
...
+internal static partial class ItemIndexRangeHelper
+{
+	internal enum IndexLocation
+	{
+		Before,
+		After,
+		Inside
+	}
+
+	internal struct Range //TODO:MZ:WinUI uses struct here
```
Comment (agneszitte) [inline src/Uno.UI/UI/Xaml/Data/ItemIndexRange/ItemIndexRangeHelper.h.mux.cs:16]: Maybe add a link in this comment to the related item or issue for this TODO

File: src/Uno.UI/UI/Xaml/Data/ItemIndexRange/ItemIndexRangeHelper.mux.cs [hunk 4]
```diff
@@ -0,0 +1,607 @@
+﻿// MUX Reference ItemIndexRangeHelper.cpp
+
+using System.Collections.Generic;
+
+namespace DirectUI.Components;
+
+internal static partial class ItemIndexRangeHelper
+{
+	#region helper functions
+	// used to check if an index is within range defined by a start and an end
```
Comment (agneszitte) [inline src/Uno.UI/UI/Xaml/Data/ItemIndexRange/ItemIndexRangeHelper.mux.cs:10]:
```suggestion
	// used to check if an index is within a range defined by a start and an end
```

File: src/Uno.UI/UI/Xaml/Data/ItemIndexRange/ItemIndexRangeHelper.mux.cs [hunk 5]
```diff
@@ -0,0 +1,607 @@
...
+				Range newRange = new(firstIndex, length);
+
+				// insert new range to selected ranges and append it to the added ranges
+				m_selectedRanges.Insert(firstRangeIndex, newRange);
+				addedRanges.Add(newRange);
+
+				return;
+			}
+
+			// probability of new range being completely inside an existing a range
```
Comment (agneszitte) [inline src/Uno.UI/UI/Xaml/Data/ItemIndexRange/ItemIndexRangeHelper.mux.cs:225]:
```suggestion
			// probability of new range being completely inside an existing range
```

File: src/Uno.UI/UI/Xaml/Data/ItemIndexRange/ItemIndexRangeHelper.mux.cs [hunk 6]
```diff
@@ -0,0 +1,607 @@
...
+
+			// find the last index' location
+			FindIndexInSelectedRanges(index, out rangeIndex, out indexInside);
+
+			pRangeIndex = rangeIndex;
+
+			if (front)
+			{
+				// special case when the returned range index is -1, this means that index is either before the first range or the ranges list is empty
+				// no need to do the other tests
```
Comment (agneszitte) [inline src/Uno.UI/UI/Xaml/Data/ItemIndexRange/ItemIndexRangeHelper.mux.cs:481]:
```suggestion
				// particular case when the returned range index is -1, this means that the index is either before the first range or the ranges list is empty
				// no need to do the other tests
```

File: src/Uno.UI/UI/Xaml/Data/ItemIndexRange/ItemIndexRangeHelper.mux.cs [hunk 7]
```diff
@@ -0,0 +1,607 @@
...
+				{
+					m_selectedRanges.Insert(currentRangeIndex + 1, new Range(lastIndex + 1, (uint)updatedLength2));
+				}
+			}
+
+			// append to removed ranges
+			removedRanges.Add(new Range(firstIndex, length));
+		}
+
+		// a helper if the deselected range is interesecting with the current range from the front or end
```
Comment (agneszitte) [inline src/Uno.UI/UI/Xaml/Data/ItemIndexRange/ItemIndexRangeHelper.mux.cs:554]:
```suggestion
		// a helper if the deselected range is intersecting with the current range from the front or end
```

File: src/Uno.UI/UI/Xaml/Data/ItemIndexRange/ItemIndexRangeHelper.h.mux.cs (hunk 3 above)
Comment (MartinZikmund) [reply src/Uno.UI/UI/Xaml/Data/ItemIndexRange/ItemIndexRangeHelper.h.mux.cs]: The todo does not apply anymore, as we are now using struct as well

**Review by MartinZikmund — APPROVED** [review]

**Pull request merged**